### DISCUSSION TITLE
Simplifying the CF firmware

### DISCUSSION THREAD

**Discussion by NageshEranki** [bitcraze/discussions#1171] [Q&A]
Hi everyone!

I'd like to make some major modifications to the Crazyflie firmware. I understand that it relies on the FreeRTOS library to allocate call stacks/queues to all [sorts of tasks.](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/) I want to make the code that runs on the CF smaller and smaller, so I want to experiment with removing multiple non-essential system tasks or replacing them with my versions. 

The example in the docs to add new system tasks is useful; but the there are simply so many tasks the MCU jumps in between that I can't keep track of which task it is currently running and for how long. Any recommendations on getting my hands dirty with the various subsytems?

For example, following the [app-channel example](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/app_layer/) let's say that the CF has only two tasks defined (and nothing else!):

1. Listen to user input for two floats.
2. Send back the sum to user.

Is there a way to make the resulting binary absolutely razor-thin and minimal? And once I'm able to, how can I verify the time taken for each task?

Thanks
Nagesh

**Comment by ataffanel** [comment]
I am not sure what you are exactly after. It is very likely you could make the firmware much simpler, mostly if you only target one platform. If you do not want to be able to fly you can even start from 0, setting up the Crazyflie pinout in STM32 Cube is quite easy and as long as you link it at the right address, you can program your firmware using the radio bootloader. You will have to port everything you want to use though.

How and what to do mostly depends of the reason: if you want to be able to analyze time spent in different tasks there is FreeRTOS runtime stats and tracing facilities that can be setup to get some measurements of tasks runtime.

Your example of the app channel would be using a single task anyway: a minimal implementation would be to receive, calculate and send the floats in a loop in a single task.

So to answer your questions:
 - To get a minimal binay size you likely have to restart from 0 adding needed pieces one by one.
 - Verifying the time taken for each tasks can be achieve using FreeRTOS [runtime stats](https://www.freertos.org/rtos-run-time-stats.html) or tracing facilities, it does not require changing the firmware size in anyway.
